homeassistant/components/sensor/youless.py

```diff
@@ -0,0 +1,138 @@
+"""
+Support for YouLess device.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.youless/
+"""
+import json
+from datetime import timedelta
+from urllib.request import urlopen
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import CONF_HOST, CONF_MONITORED_CONDITIONS
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+SENSOR_POWER = 'pwr'
+SENSOR_POWER_NET = 'net'
+SENSOR_P_LOW = 'p1'
+SENSOR_P_HIGH = 'p2'
+SENSOR_DELIVERY_LOW = 'n1'
+SENSOR_DELIVERY_HIGH = 'n2'
+SENSOR_METER_EXTRA = 'cs0'
+SENSOR_USAGE_EXTRA = 'ps0'
+SENSOR_GAS = 'gas'
+
+SENSOR_TYPES = {
+    SENSOR_POWER: ['Usage', 'usage', 'W', 'mdi:flash'],
+    SENSOR_POWER_NET: ['Power', 'meter', 'W', 'mdi:gauge'],
+    SENSOR_P_LOW: ['Power Low', 'meter_low', 'kWh', 'mdi:flash'],
+    SENSOR_P_HIGH: ['Power High', 'meter_high', 'kWh', 'mdi:flash'],
+    SENSOR_DELIVERY_LOW: ['Power Delivery Low', 'delivery_low', 'kWh',
+                          'mdi:gauge'],
+    SENSOR_DELIVERY_HIGH: ['Power Delivery High', 'delivery_high', 'kWh',
+                           'mdi:gauge'],
+    SENSOR_METER_EXTRA: ['Power Meter Extra', 'meter_extra', 'kWh',
+                         'mdi:gauge'],
+    SENSOR_USAGE_EXTRA: ['Power usage Extra', 'current_usage_extra', 'W',
+                         'mdi:flash'],
+    SENSOR_GAS: ['Gas', 'gas', 'm3', 'mdi:fire']
+}
+
+CONFIG_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_HOST): cv.string,
+    vol.Optional(CONF_MONITORED_CONDITIONS,
+                 default=[SENSOR_POWER, SENSOR_POWER_NET]): vol.All(
+                     cv.ensure_list, vol.Length(min=1),
+                     [vol.In(SENSOR_TYPES)])
+})
+
+SENSOR_PREFIX = 'youless_'
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the platform based on the configuration."""
+    host = config.get(CONF_HOST)
+    sensors = config.get(CONF_MONITORED_CONDITIONS)
+    data_bridge = YoulessDataBridge(host)
+
+    devices = []
+    for sensor in sensors:
+        sensor_config = SENSOR_TYPES[sensor]
+        devices.append(YoulessSensor(data_bridge, sensor_config[0], sensor,
+                                     sensor_config[1], sensor_config[2],
+                                     sensor_config[3]))
+
+    add_devices(devices)
+
+
+class YoulessDataBridge:
```

Comment:
Technically, this class runs afoul of the Home Assistant rule that protocol handling for the device must be in an external PyPI package. However, I am unsure if this is so tiny that it is acceptable to bend the rule here.

I will defer to @balloob.

Reply:
Yes, please extract this to standalone library published on pypi:

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

Reply:
> Yes, please extract this to standalone library published on pypi:
> 
> https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

Unfortunately it seems that this requirement has killed the project, can you help is to get this done?